# Лабораторная работа №3

Выполнил студент группы 9381 Птичкин Сергей.

##### Вариант №14

### Цель работы:
Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

### Задачи лабораторной работы
1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

### 1. Решение диофантова уравнения.

5415x + 4769y = -152

In [3]:
from sympy.solvers.diophantine import diophantine
from sympy import symbols

x,y,z = symbols('x y z')

diophantine(5415*x + 4769*y + 152)

{(251*t_0 - 768, 872 - 285*t_0)}

Ответ: x = -768, y = 872

### 2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 7,m = 39).
Шифрованное сообщение имеет вид (33; 18; 14; 11). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “ё” и “ъ”).

In [5]:
from sympy.crypto.crypto import rsa_private_key, decipher_rsa

code = [33, 18, 14, 11]
key = rsa_private_key(39,7)
for i in code:
    print(decipher_rsa(i, key), end=" ")

6 21 14 2 

Ответ: Д У М А

### 3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

f(x, y, z) = ((zx) xor (yz)) ∨ ((x ∨ z) xor y)

In [7]:
from sympy.logic.boolalg import to_cnf, to_dnf

print("ДНФ: ", to_dnf(((z&x)^(y&z))|((x|z)^y), simplify=True))
print("КНФ: ", to_cnf(((z&x)^(y&z))|((x|z)^y), simplify=True))

ДНФ:  (x & ~y) | (y & ~x) | (z & ~x)
КНФ:  (x | y | z) & (~x | ~y)


### Вывод
В процессе выполнения лабораторной работы были изучены функции библиотеки SymPy Языка Программирования Python для решения задач дискретной математики и математической логики.